evafast.lua

```diff
@@ -43,244 +43,243 @@ local options = {
+    speedup = true
+    speed_timer.timeout = 0
+    speed_timer:resume()
+    speed_timer.timeout = options.speed_interval
 end
 
-local function evafast(keypress)
-    if jumps_reset_speed and not toggle_state and (keypress["event"] == "up" or keypress["event"] == "press") then
-        speedup = false
-        speedup_target = nil
+local function evafast_slowdown()
+    toggled_display = false
+    toggled = false
+    speedup = false
+end
+
+local function evafast_toggle()
+    if speedup and speed_timer:is_enabled() then
+        evafast_slowdown()
+    else
+        evafast_speedup()
     end
+end
 
-    if options.seek_distance == 0 then
-        if keypress["event"] == "up" or keypress["event"] == "press" then
-            speedup = false
-            no_speedup = true
-            repeated = false
-            speedup_target = nil
+local function adjust_speed()
+    local current_time = mp.get_property_number("time-pos", 0)
+    local current_speed = mp.get_property_number("speed", 1)
+    local target_speed = original_speed
+
+    if speedup then
+        target_speed = options.speed_cap
+
+        if has_subtitle and options.subs_speed_cap ~= options.speed_cap then
+            if mp.get_property("sub-start") ~= nil then
+                target_speed = options.subs_speed_cap
+            elseif options.subs_lookahead then
+                if next_sub_at < current_time then
+                    next_sub_at = next_sub(current_time)
+                end
+                if next_sub_at >= current_time then
+                    local time_for_correction = speed_transition(current_speed, options.subs_speed_cap)
+                    if (time_for_correction * current_speed + current_time) > next_sub_at then
+                        target_speed = options.subs_speed_cap
+                    end
+                end
+            end
         end
-        if keypress["event"] == "down" then
-            keypress["event"] = "repeat"
+
+        if speedup_target ~= nil then
+            if current_time >= speedup_target then
+                evafast_slowdown()
+            else
+                local time_for_correction = speed_transition(current_speed, original_speed)
+                if (time_for_correction * current_speed + current_time) > speedup_target then
+                    evafast_slowdown()
+                    target_speed = original_speed
+                end
+            end
         end
     end
 
-    if keypress["event"] == "up" or keypress["event"] == "press" then
-        toggle_display = toggle_state
-        if toggle_state and jumps_reset_speed then
-            speedup = false
-            speedup_target = nil
-        end
-        if speed_timer ~= nil and not toggle_state and mp.get_property_number("speed") == 1 and ((not options.subs_speed_cap or mp.get_property("sub-start") == nil) and options.speed_cap or options.subs_speed_cap) ~= 1 then
+    if math.floor(target_speed * 1000) == math.floor(current_speed * 1000) then
```

Comment:
I hate this, but simple comparison of floats that should be the same doesn't work here.  
Look for a better solution, or benchmark this vs `math.abs(target_speed - current_speed) < 0.00001` vs `string.format()`.